mongodbatlas/fw_resource_mongodbatlas_search_deployment.go

```diff
@@ -0,0 +1,368 @@
+	clusterName := searchDeploymentPlan.ClusterName.ValueString()
+	searchDeploymentReq := newSearchDeploymentReq(ctx, &searchDeploymentPlan)
+	if _, _, err := connV2.AtlasSearchApi.CreateAtlasSearchDeployment(ctx, projectID, clusterName, &searchDeploymentReq).Execute(); err != nil {
+		resp.Diagnostics.AddError("error during search deployment creation", err.Error())
+		return
+	}
+
+	createTimeout, diags := searchDeploymentPlan.Timeouts.Create(ctx, defaultSearchNodeTimeout)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	deploymentResp, err := waitSearchNodeStateTransition(ctx, projectID, clusterName, connV2, createTimeout)
+	if err != nil {
+		resp.Diagnostics.AddError("error during search deployment creation", err.Error())
+		return
+	}
+	newSearchNodeModel, diagnostics := newTFSearchDeployment(ctx, clusterName, deploymentResp, &searchDeploymentPlan.Timeouts)
+	resp.Diagnostics.Append(diagnostics...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, newSearchNodeModel)...)
+}
+
+func (r *SearchDeploymentRS) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var searchDeploymentPlan tfSearchDeploymentRSModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &searchDeploymentPlan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	connV2 := r.client.AtlasV2
+	projectID := searchDeploymentPlan.ProjectID.ValueString()
+	clusterName := searchDeploymentPlan.ClusterName.ValueString()
+	deploymentResp, _, err := connV2.AtlasSearchApi.GetAtlasSearchDeployment(ctx, projectID, clusterName).Execute()
+	if err != nil {
+		resp.Diagnostics.AddError("error getting search deployment information", err.Error())
+		return
+	}
+
+	newSearchNodeModel, diagnostics := newTFSearchDeployment(ctx, clusterName, deploymentResp, &searchDeploymentPlan.Timeouts)
+	resp.Diagnostics.Append(diagnostics...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, newSearchNodeModel)...)
+}
+
+func (r *SearchDeploymentRS) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var searchDeploymentPlan tfSearchDeploymentRSModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &searchDeploymentPlan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	connV2 := r.client.AtlasV2
+	projectID := searchDeploymentPlan.ProjectID.ValueString()
+	clusterName := searchDeploymentPlan.ClusterName.ValueString()
+	searchDeploymentReq := newSearchDeploymentReq(ctx, &searchDeploymentPlan)
+	if _, _, err := connV2.AtlasSearchApi.UpdateAtlasSearchDeployment(ctx, projectID, clusterName, &searchDeploymentReq).Execute(); err != nil {
+		resp.Diagnostics.AddError("error during search deployment update", err.Error())
+		return
+	}
+
+	updateTimeout, diags := searchDeploymentPlan.Timeouts.Update(ctx, defaultSearchNodeTimeout)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
```

Comment:
I see this is more or less the same that is in `create` any chance we could make this method leaner and move this inside the `waitSearch...`? or just make it a separate method with a one-liner call here..

Reply:
So there are some details here to take into account:
- for the simple approach of extracting to separate methods and having a one-liner: the part checking if there are errors and finish the update function (L193-L195) can not be extracted. L192 could potentially be extracted into separate function, but not in favour of passing resp and mutating within another function (harder to follow).
- for having the logic inside waitSearch function, an additional parameter would be added to receive the function `searchDeploymentPlan.Timeouts.Create` and `searchDeploymentPlan.Timeouts.Update` which add some complexity to the method signature. This is why I end up passing `timeout time.Duration` directly with some common code in each method.

Reply:
moving forward with the merge, if any further action comes from this comment we can address in a followup PR.